File: 프로그래머스/2/169198. 당구 연습/당구 연습.cpp
#include <string>
#include <vector>
#include <climits>

using namespace std;

int sq(int v) { return v * v; }

vector<int> solution(int m, int n, int startX, int startY, vector<vector<int>> balls) {
    vector<int> answer;
    
    for (auto& ball : balls) {
        int x = ball[0], y = ball[1];
        int best = INT_MAX;
     
        // 왼쪽
        if (!(startY == y && startX > x)) {
            int mx = -x, my = y;
            best = min(best, sq(startX - mx) + sq(startY - my));
        }
        
        // 오른쪽
        if (!(startY == y && startX < x)) {
            int mx = 2 * m - x, my = y;
            best = min(best, sq(startX - mx) + sq(startY - my));
        }
        
        // 아래
        if (!(startX == x && startY > y)) {
            int mx = x, my = -y;
            best = min(best, sq(startX - mx) + sq(startY - my));
        }
        
        // 위
        if (!(startX == x && startY < y)) {
            int mx = x, my = 2 * n - y;
            best = min(best, sq(startX - mx) + sq(startY - my));
        }
        
        answer.push_back(best);
    }
    return answer;
}

File: 프로그래머스/2/169198. 당구 연습/당구 연습.java
class Solution {
    public int[] solution(int m, int n, int startX, int startY, int[][] balls) {
        int[] answer = new int[balls.length];
        for (int i = 0; i < balls.length; i++) {
            int targetX = balls[i][0];
            int targetY = balls[i][1];
            
            int res = Integer.MAX_VALUE;
            
            // 위
            if (!(startX == targetX && targetY < startY)) {
                res = Math.min(res, dist(startX, startY, targetX, -targetY));
            }
            
            // 아래
            if (!(startX == targetX && targetY > startY)) {
                res = Math.min(res, dist(startX, startY, targetX, 2 * n - targetY));
            }
            
            // 왼
            if (!(startX > targetX && targetY == startY)) {
                res = Math.min(res, dist(startX, startY, -targetX, targetY));
            }
            
            // 오
            if (!(startX < targetX && targetY == startY)) {
                res = Math.min(res, dist(startX, startY, 2 * m - targetX, targetY));
            }
            answer[i] = res;
        }
        return answer;
    }
    
    private int dist(int x1, int y1, int x2, int y2) {
        int dx = x1 - x2;
        int dy = y1 - y2;
        return dx * dx + dy * dy;
    }
}